// vscode-chrome-debug-core/out/src/index.js
"use strict";
/*---------------------------------------------------------
 * Copyright (C) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
/** Normally, a consumer could require and use this and get the same instance. But if -core is npm linked, there may be two instances of file in play. */
const vscode_debugadapter_1 = require("vscode-debugadapter");
exports.logger = vscode_debugadapter_1.logger;
const chromeConnection = require("./chrome/chromeConnection");
exports.chromeConnection = chromeConnection;
const chromeDebugAdapter_1 = require("./chrome/chromeDebugAdapter");
exports.ChromeDebugAdapter = chromeDebugAdapter_1.ChromeDebugAdapter;
const chromeDebugSession_1 = require("./chrome/chromeDebugSession");
exports.ChromeDebugSession = chromeDebugSession_1.ChromeDebugSession;
const chromeTargetDiscoveryStrategy = require("./chrome/chromeTargetDiscoveryStrategy");
exports.chromeTargetDiscoveryStrategy = chromeTargetDiscoveryStrategy;
const chromeUtils = require("./chrome/chromeUtils");
exports.chromeUtils = chromeUtils;
const stoppedEvent = require("./chrome/stoppedEvent");
exports.stoppedEvent = stoppedEvent;
const internalSourceBreakpoint_1 = require("./chrome/internalSourceBreakpoint");
exports.InternalSourceBreakpoint = internalSourceBreakpoint_1.InternalSourceBreakpoint;
const errors_1 = require("./errors");
exports.ErrorWithMessage = errors_1.ErrorWithMessage;
const basePathTransformer_1 = require("./transformers/basePathTransformer");
exports.BasePathTransformer = basePathTransformer_1.BasePathTransformer;
const urlPathTransformer_1 = require("./transformers/urlPathTransformer");
exports.UrlPathTransformer = urlPathTransformer_1.UrlPathTransformer;
const lineNumberTransformer_1 = require("./transformers/lineNumberTransformer");
exports.LineColTransformer = lineNumberTransformer_1.LineColTransformer;
const baseSourceMapTransformer_1 = require("./transformers/baseSourceMapTransformer");
exports.BaseSourceMapTransformer = baseSourceMapTransformer_1.BaseSourceMapTransformer;
const utils = require("./utils");
exports.utils = utils;
const telemetry = require("./telemetry");
exports.telemetry = telemetry;
const variables = require("./chrome/variables");
exports.variables = variables;
const nullLogger_1 = require("./nullLogger");
exports.NullLogger = nullLogger_1.NullLogger;
const executionTimingsReporter = require("./executionTimingsReporter");
exports.executionTimingsReporter = executionTimingsReporter;
const chromeTargetDiscoveryStrategy_1 = require("./chrome/chromeTargetDiscoveryStrategy");
exports.Version = chromeTargetDiscoveryStrategy_1.Version;
exports.TargetVersions = chromeTargetDiscoveryStrategy_1.TargetVersions;
const breakpoints_1 = require("./chrome/breakpoints");
exports.Breakpoints = breakpoints_1.Breakpoints;
const scripts_1 = require("./chrome/scripts");
exports.ScriptContainer = scripts_1.ScriptContainer;
